docs(sources/fluent-bit): validate and promote Fluent Bit → GlassFlow OTLP (ETL-1132) (#762)
glassflow/clickhouse-etl: docs/app/sources/fluent-bit/page.mdx

---
title: 'Fluent Bit'

> description: 'Use Fluent Bit to collect application logs and metrics from your systems and ship them to ClickHouse via the GlassFlow OTLP receiver'

Reviewer: ```suggestion
description: 'Use Fluent Bit to collect application logs and metrics from your systems and ship them to ClickHouse via the GlassFlow OTLP receiver.'
```

</Callout>

The GlassFlow pipeline that writes these records to ClickHouse uses an `otlp.logs` source. For the sink `mapping` (OTel field → ClickHouse column) and the full set of mappable fields, see [OTLP — Examples](/sources/otlp/examples) and the [OTLP Schema Reference](/sources/otlp/schema-reference#otlp-logs-fields).

## Metrics

Whatever metrics Fluent Bit collects, the `opentelemetry` output forwards them to GlassFlow the same way. The example below scrapes an application's Prometheus endpoint, but the output config is identical regardless of which metric input you use:

```ini
[INPUT]
    Name             prometheus_scrape
    Tag              app.metrics
    Host             my-app
    Port             9111
    Metrics_path     /metrics
    Scrape_interval  10s

[OUTPUT]
    Name          opentelemetry
    Match         app.metrics
    Host          <glassflow-otlp-receiver-host>
    Port          4318
    Metrics_uri   /v1/metrics
    Tls           off
    Header        x-glassflow-pipeline-id <your-metrics-pipeline-id>
    Retry_Limit   3
```

> Counters, gauges, and histograms arrive in the pipeline's target table with their `metric_type` correctly identified. Any Prometheus labels are preserved as attributes — in the sample below only the counter carries labels; the gauge and histogram have none:

Reviewer: ```suggestion
Counters, gauges, and histograms arrive in the pipeline's target table with their `metric_type` correctly identified. Any Prometheus labels are preserved as attributes. In the sample below, only the counter carries labels; the gauge and histogram have none:
```

| `metric_name` | `metric_type` | `attributes` |
|---|---|---|
| `myapp_http_requests_total` | `sum` | `{'method':'GET','route':'/api/orders','status':'200'}` |
| `myapp_inflight_requests` | `gauge` | `{}` |
| `myapp_request_duration_seconds` | `histogram` | `{}` |

Metric descriptions and units are preserved alongside the values.

## Traces

> Fluent Bit can act as a trace forwarder — apps send OTLP traces to its `opentelemetry` input, and the `opentelemetry` output forwards them to GlassFlow. Two requirements apply:

Reviewer: ```suggestion
Fluent Bit can act as a trace forwarder. Applications send OTLP traces to its `opentelemetry` input, and the `opentelemetry` output forwards them to GlassFlow. Two requirements apply:
```

```

Spans arrive in the pipeline's target table with their parent/child structure, attributes, and 16-byte trace IDs intact.

If your applications can only emit OTLP/HTTP traces, or you are pinned to Fluent Bit < 5.0, forward traces with a dedicated OpenTelemetry Collector instead — see [OTLP — Sending Data](/sources/otlp/sending-data).

## Verifying data flow

Three places to look:

1. **Fluent Bit's own logs.** A healthy output looks like `[output:opentelemetry:opentelemetry.0] <host>:4318, HTTP status=200`. Any non-2xx response is logged at `[error]` and the chunk is queued for retry.
2. **The GlassFlow OTLP receiver's metrics.** `glassflow_gfm_bytes_processed_total{component="otlp.logs", direction="in", pipeline_id="<your-pipeline-id>"}` increments on every accepted batch. If `direction="in"` rises but `direction="out"` stays flat, the receiver accepted the records but couldn't publish them to NATS — check the pipeline's `/api/v1/pipeline/<id>/health` endpoint.
3. **ClickHouse.** Recent rows appear in the table configured in your pipeline's sink; filter by an attribute your data carries to confirm records are arriving.

## Caveats

> - **Misconfigured pipeline IDs fail quietly.** Fluent Bit's `opentelemetry` output treats every non-2xx response as retryable, including 4xx. If the `Header x-glassflow-pipeline-id` value is wrong or the pipeline doesn't exist, Fluent Bit retries each batch up to `Retry_Limit` and then drops it — there's no fatal error to alert on. Watch Fluent Bit's `fluentbit_output_retries_total` for that output — a steadily-climbing counter with no successes is the canary.

Reviewer: ```suggestion
- **Misconfigured pipeline IDs fail quietly.** Fluent Bit's `opentelemetry` output treats every non-2xx response as retryable, including 4xx. If the `Header x-glassflow-pipeline-id` value is wrong or the pipeline doesn't exist, Fluent Bit retries each batch up to `Retry_Limit` and then drops it. There's no fatal error to alert on. Watch Fluent Bit's `fluentbit_output_retries_total` for that output. A steadily-climbing counter with no successes is the canary.
```
